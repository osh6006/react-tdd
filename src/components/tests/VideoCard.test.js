import { render, screen } from "@testing-library/react";
import useEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";

import { Route, useLocation } from "react-router-dom";

import VideoCard from "../VideoCard";

import { formatAgo } from "../../util/date";
import { fakeVideo as video } from "../../tests/videos";
import { withRouter } from "../../tests/utils";

describe("VideoCard", () => {
  const { title, channelTitle, publishedAt, thumbnails } = video.snippet;

  it(`renders grid type`, () => {
    const component = renderer.create(
      withRouter(<Route path="/" element=<VideoCard video={video} /> />)
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it(`renders list type`, () => {
    const component = renderer.create(
      withRouter(
        <Route path="/" element=<VideoCard video={video} type={"list"} /> />
      )
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders video item", () => {
    render(
      withRouter(<Route path="/" element={<VideoCard video={video} />} />)
    );

    const image = screen.getByRole("img");

    expect(image.src).toBe(thumbnails.medium.url);
    expect(image.alt).toBe(title);
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(channelTitle)).toBeInTheDocument();
    expect(screen.getByText(formatAgo(publishedAt))).toBeInTheDocument();
  });

  it("navigates to detailed video page with video state when clicked", () => {
    function LocationStateDisplay() {
      return <pre>{JSON.stringify(useLocation().state)}</pre>;
    }

    render(
      withRouter(
        <>
          <Route path="/" element={<VideoCard video={video} />} />
          <Route
            path={`/videos/watch/${video.id}`}
            element={<LocationStateDisplay />}
          />
        </>
      )
    );

    const card = screen.getByRole("listitem");
    useEvent.click(card);

    expect(screen.getByText(JSON.stringify({ video }))).toBeInTheDocument();
  });
});
