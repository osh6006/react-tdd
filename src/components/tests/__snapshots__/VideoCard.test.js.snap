// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoCard renders grid type 1`] = `
<li
  className=""
  onClick={[Function]}
>
  <img
    alt="GloRilla, Cardi B - Tomorrow 2 (Official Music Video)"
    className="w-full"
    src="https://i.ytimg.com/vi/zwa7NzNBQig/mqdefault.jpg"
  />
  <div>
    <p
      className="font-semibold my-2 line-clamp-2"
    >
      GloRilla, Cardi B - Tomorrow 2 (Official Music Video)
    </p>
    <p
      className="text-sm opacity-80"
    >
      GloRillaVEVO
    </p>
    <p
      className="text-sm opacity-80"
    >
      1 year ago
    </p>
  </div>
</li>
`;

exports[`VideoCard renders list type 1`] = `
<li
  className="flex gap-1 m-2"
  onClick={[Function]}
>
  <img
    alt="GloRilla, Cardi B - Tomorrow 2 (Official Music Video)"
    className="w-60 mr-2"
    src="https://i.ytimg.com/vi/zwa7NzNBQig/mqdefault.jpg"
  />
  <div>
    <p
      className="font-semibold my-2 line-clamp-2"
    >
      GloRilla, Cardi B - Tomorrow 2 (Official Music Video)
    </p>
    <p
      className="text-sm opacity-80"
    >
      GloRillaVEVO
    </p>
    <p
      className="text-sm opacity-80"
    >
      1 year ago
    </p>
  </div>
</li>
`;
